src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIChatClient.cs

```diff
@@ -370,6 +344,50 @@ private sealed class FunctionCallInfo
         public StringBuilder? Arguments;
     }
 
+    private static UsageDetails ToUsageDetails(ChatTokenUsage tokenUsage)
+    {
+        var destination = new UsageDetails
+        {
+            InputTokenCount = tokenUsage.InputTokenCount,
+            OutputTokenCount = tokenUsage.OutputTokenCount,
+            TotalTokenCount = tokenUsage.TotalTokenCount,
+            AdditionalCounts = new(),
+        };
+
+        if (tokenUsage.InputTokenDetails is ChatInputTokenUsageDetails inputDetails)
+        {
+            if (inputDetails.AudioTokenCount is int audioTokenCount)
+            {
+                destination.AdditionalCounts.Add(
+                    $"{nameof(ChatTokenUsage.InputTokenDetails)}.{nameof(ChatInputTokenUsageDetails.AudioTokenCount)}",
+                    audioTokenCount);
+            }
+
+            if (inputDetails.CachedTokenCount is int cachedTokenCount)
+            {
+                destination.AdditionalCounts.Add(
+                    $"{nameof(ChatTokenUsage.InputTokenDetails)}.{nameof(ChatInputTokenUsageDetails.CachedTokenCount)}",
+                    cachedTokenCount);
+            }
+        }
+
+        if (tokenUsage.OutputTokenDetails is ChatOutputTokenUsageDetails outputDetails)
+        {
+            if (outputDetails.AudioTokenCount is int audioTokenCount)
+            {
+                destination.AdditionalCounts.Add(
+                    $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.AudioTokenCount)}",
+                    audioTokenCount);
+            }
+
+            destination.AdditionalCounts.Add(
+                $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.ReasoningTokenCount)}",
+                outputDetails.ReasoningTokenCount);
+        }
+
+        return destination;
+    }
+
```

Comment:
Were you also going to update the ollama impl to move the timings to usage details?

Reply:
Note that I did *not* add a `_ns` suffix to the keys, based on our separate discussion.